### PR TITLE
Add the admin functionality

### PR DESCRIPTION
**Pull request by antriksh-webonise**
- Add the functionality to add,delete,upadte and view all users
- Add the functionality to view all the bookings

### CODE REVIEW

File: src/main/java/com/webonise/rbs/RbsApplication.java [hunk 1]
```diff
@@ -1,12 +1,17 @@
 package com.webonise.rbs;
 
+import com.webonise.rbs.service.UserService;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.ComponentScan;
 
 @SpringBootApplication
+/*@ComponentScan("com.webonise.rbs")*/
```
Comment (akshaym-webonise) [inline src/main/java/com/webonise/rbs/RbsApplication.java]: Please do not push commented code.

Comment (himani-webonise) [inline src/main/java/com/webonise/rbs/RbsApplication.java]: Remove commented code

File: src/main/java/com/webonise/rbs/controller/UserController.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    @GetMapping(value = "/",produces = MediaType.TEXT_HTML_VALUE)
+    public String savePage(Model model) {
+        model.addAttribute("user", new User());
+        model.addAttribute("allUsers", userService.getAllUsers());
+        return "users";
+    }
+
+    @PostMapping("/save")
+    public String saveUser(@ModelAttribute("user") User user, final RedirectAttributes redirectAttributes) {
```
Comment (akshaym-webonise) [inline src/main/java/com/webonise/rbs/controller/UserController.java]: Follow proper REST conventions

File: pom.xml [hunk 3]
```diff
@@ -63,9 +58,16 @@
 			<scope>test</scope>
 		</dependency>
 		<dependency>
-			<groupId>org.springframework.security</groupId>
-			<artifactId>spring-security-test</artifactId>
-			<scope>test</scope>
+			<groupId>net.sourceforge.nekohtml</groupId>
+			<artifactId>nekohtml</artifactId>
```
Comment (pareekshita-webonise) [inline pom.xml]: what is the use of this dependency?

File: pom.xml [hunk 4]
```diff
@@ -63,9 +58,16 @@
...
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-actuator</artifactId>
```
Comment (pareekshita-webonise) [inline pom.xml]: same as above.

File: src/main/java/com/webonise/rbs/entity/User.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    public Long getRoleId() {
+        return roleId;
+    }
+
+    public void setRoleId(Long roleId) {
+        this.roleId = roleId;
+    }
+}
+
```
Comment (akshaym-webonise) [inline src/main/java/com/webonise/rbs/entity/User.java]: Remove the new lines. Follow the same code format guidelines throughout the code.

File: src/main/java/com/webonise/rbs/service/UserServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+import java.util.Collection;
+import java.util.List;
+
+@Service
+@Transactional
+public class UserServiceImpl implements UserService {
+
+    @Autowired
+    UserRepository userRepository;
```
Comment (akshaym-webonise) [inline src/main/java/com/webonise/rbs/service/UserServiceImpl.java:16]: make **userRepository** private

File: src/main/java/com/webonise/rbs/controller/UserController.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.servlet.ModelAndView;
+import org.springframework.web.servlet.mvc.support.RedirectAttributes;
+
+import java.util.ArrayList;
+
+@Controller
+@RequestMapping("/user")
```
Comment (himani-webonise) [inline src/main/java/com/webonise/rbs/controller/UserController.java]: Make it "/users"
http://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api

File: src/main/java/com/webonise/rbs/controller/UserController.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+public class UserController {
+
+    @Autowired
+    UserServiceImpl userService;
```
Comment (himani-webonise) [inline src/main/java/com/webonise/rbs/controller/UserController.java]: Please make it private

Comment (himani-webonise) [reply src/main/java/com/webonise/rbs/controller/UserController.java]: Autowire interface 'UserService' here
https://stackoverflow.com/questions/12899372/spring-why-do-we-autowire-the-interface-and-not-the-implemented-class

File: src/main/resources/application.properties [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.MySQLDialect
+spring.jpa.properties.hibernate.id.new_generator_mappings = false
+spring.jpa.properties.hibernate.format_sql = true
+
+spring.thymeleaf.prefix = classpath:/templates/
+spring.thymeleaf.suffix = .html
+
+# Spring JPA
+#spring.jpa.databasePlatform=org.hibernate.dialect.MySQLDialect
+#spring.jpa.hibernate.ddl-auto=create-drop
```
Comment (akshaym-webonise) [inline src/main/resources/application.properties]: again commented code.

File: src/main/java/com/webonise/rbs/controller/UserController.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+    @GetMapping(value = "/",produces = MediaType.TEXT_HTML_VALUE)
+    public String savePage(Model model) {
+        model.addAttribute("user", new User());
+        model.addAttribute("allUsers", userService.getAllUsers());
+        return "users";
+    }
+
+    @PostMapping("/save")
```
Comment (himani-webonise) [inline src/main/java/com/webonise/rbs/controller/UserController.java]: Mapping type Post signifies that you are saving something, so need to have uri /save.
https://hackernoon.com/restful-api-designing-guidelines-the-best-practices-60e1d954e7c9

File: src/main/resources/templates/editUsers.html [hunk 11]
```diff
@@ -0,0 +1,43 @@
+<!DOCTYPE html>
+<html xmlns="http://www.w3.org/1999/xhtml" xmlns:th="http://www.thymeleaf.org" xmlns:sec="http://www.thymeleaf.org/thymeleaf-extras-springsecurity3">
+<head>
+    <title>Edit Employee</title>
+</head>
+<body>
+    <div class="container">
+        <div class="row">
+            <div class="col-md-offset-3 col-md-6">
+                <div th:if="${saveUser eq 'unsuccess'}">
```
Comment (pareekshita-webonise) [inline src/main/resources/templates/editUsers.html]: what is 'unsuccess'. Make it more meaningful and appropriate

File: src/main/java/com/webonise/rbs/controller/UserController.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        } else {
+            redirectAttributes.addFlashAttribute("saveUser", "unsuccess");
+        }
+        return "redirect:/user/";
+    }
+
+    @GetMapping(value = "/{operation}/{id}")
+    public String editRemoveUser(@PathVariable("operation") String operation, @PathVariable("id") Long id, final RedirectAttributes redirectAttributes, Model model) {
+
+        if(operation.equals("delete")) {
```
Comment (akshaym-webonise) [inline src/main/java/com/webonise/rbs/controller/UserController.java]: wrong implementation for
1. Delete user
2. Retrieve user by id
These two should be 2 different tasks.

File: src/main/java/com/webonise/rbs/controller/UserController.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+        if(userService.addUser(user)!=null) {
+            redirectAttributes.addFlashAttribute("saveUser", "success");
+        } else {
+            redirectAttributes.addFlashAttribute("saveUser", "unsuccess");
+        }
+        return "redirect:/user/";
+    }
+
+    @GetMapping(value = "/{operation}/{id}")
+    public String editRemoveUser(@PathVariable("operation") String operation, @PathVariable("id") Long id, final RedirectAttributes redirectAttributes, Model model) {
```
Comment (himani-webonise) [inline src/main/java/com/webonise/rbs/controller/UserController.java]: Please create separate API endpoint each for edit i.e. PUT and remove i.e. Delete. Don't club two different behaviour here.

File: src/main/resources/application.properties (hunk 9 above)
Comment (pareekshita-webonise) [inline src/main/resources/application.properties]: why commented code ?

File: src/main/java/com/webonise/rbs/service/ViewBookingsServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+
+@Service
+@Transactional
+public class ViewBookingsServiceImpl implements ViewBookingsService {
+
+    @Autowired
+    BookingRepository bookingRepository;
```
Comment (pareekshita-webonise) [inline src/main/java/com/webonise/rbs/service/ViewBookingsServiceImpl.java]: make it private

File: src/main/java/com/webonise/rbs/controller/UserController.java (hunk 13 above)
Comment (akshaym-webonise) [inline src/main/java/com/webonise/rbs/controller/UserController.java]: Single responsibility principle voilated.

File: src/main/java/com/webonise/rbs/service/ViewBookingsServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+import com.webonise.rbs.repository.BookingRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+
+@Service
+@Transactional
+public class ViewBookingsServiceImpl implements ViewBookingsService {
```
Comment (pareekshita-webonise) [inline src/main/java/com/webonise/rbs/service/ViewBookingsServiceImpl.java]: Service should be generic. Like BookingService which will contain method to view bookings, add booking , updat  booking etc.

File: src/main/java/com/webonise/rbs/service/UserServiceImpl.java (hunk 6 above)
Comment (pareekshita-webonise) [inline src/main/java/com/webonise/rbs/service/UserServiceImpl.java]: make it private

File: src/main/java/com/webonise/rbs/controller/ViewBookingsController.java [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Controller
+@RequestMapping("/bookings")
+public class ViewBookingsController {
+
+    @Autowired
+    ViewBookingsServiceImpl viewBookingService;
```
Comment (himani-webonise) [inline src/main/java/com/webonise/rbs/controller/ViewBookingsController.java]: Same as earlier, make it private. Do this at other places as well.

File: src/main/java/com/webonise/rbs/entity/User.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+
+
```
Comment (pareekshita-webonise) [inline src/main/java/com/webonise/rbs/entity/User.java]: why so many blank lines. Keep the formatting same.

File: src/main/java/com/webonise/rbs/controller/ViewBookingsController.java (hunk 16 above)
Comment (himani-webonise) [reply src/main/java/com/webonise/rbs/controller/ViewBookingsController.java]: Don't Autowire interface do this at other places as well.

File: src/main/java/com/webonise/rbs/entity/User.java [hunk 18]
```diff
@@ -0,0 +1,84 @@
+package com.webonise.rbs.entity;
+
+import javax.persistence.*;
```
Comment (pareekshita-webonise) [inline src/main/java/com/webonise/rbs/entity/User.java]: do not import whole package. Import only whats needed.

File: src/main/java/com/webonise/rbs/controller/ViewBookingsController.java (hunk 16 above)
Comment (pareekshita-webonise) [inline src/main/java/com/webonise/rbs/controller/ViewBookingsController.java]: make it private

File: src/main/java/com/webonise/rbs/controller/UserController.java (hunk 8 above)
Comment (pareekshita-webonise) [inline src/main/java/com/webonise/rbs/controller/UserController.java]: make it private

File: src/main/java/com/webonise/rbs/entity/User.java (hunk 18 above)
Comment (himani-webonise) [inline src/main/java/com/webonise/rbs/entity/User.java]: Do specific import.

File: src/main/java/com/webonise/rbs/entity/User.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+    public Long getRoleId() {
+        return roleId;
+    }
+
+    public void setRoleId(Long roleId) {
+        this.roleId = roleId;
+    }
+}
+
+
```
Comment (himani-webonise) [inline src/main/java/com/webonise/rbs/entity/User.java]: Remove extra lines

File: src/main/java/com/webonise/rbs/repository/BookingRepository.java [hunk 20]
```diff
@@ -0,0 +1,7 @@
+package com.webonise.rbs.repository;
+
+import com.webonise.rbs.entity.Booking;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface BookingRepository extends JpaRepository<Booking, Long> {
+}
```
Comment (himani-webonise) [inline src/main/java/com/webonise/rbs/repository/BookingRepository.java]: Add new line, do this change for other class also.
https://stackoverflow.com/questions/5813311/no-newline-at-end-of-file

File: src/main/java/com/webonise/rbs/repository/UserRepository.java [hunk 21]
```diff
@@ -0,0 +1,7 @@
+package com.webonise.rbs.repository;
+
+import com.webonise.rbs.entity.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface UserRepository extends JpaRepository<User, Long> {
+}
```
Comment (himani-webonise) [inline src/main/java/com/webonise/rbs/repository/UserRepository.java:7]: Same as above

File: src/main/java/com/webonise/rbs/controller/UserController.java [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+                return "editUsers";
+            } else {
+                redirectAttributes.addFlashAttribute("status","notfound");
+            }
+        }
+
+        return "redirect:/user/";
+    }
+
+    @PostMapping(value = "/update")
```
Comment (himani-webonise) [inline src/main/java/com/webonise/rbs/controller/UserController.java]: Use @PutMapping here and no need to have uri like this /update

File: src/main/java/com/webonise/rbs/RbsApplication.java [hunk 23]
```diff
@@ -1,12 +1,16 @@
 package com.webonise.rbs;
 
+import com.webonise.rbs.service.UserService;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.ComponentScan;
```
Comment (akshaym-webonise) [inline src/main/java/com/webonise/rbs/RbsApplication.java:7]: why have we imported these new classes to RbsApplication? I dont see their usage anywhere.

File: src/main/java/com/webonise/rbs/controller/UserController.java [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.ModelAttribute;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.servlet.mvc.support.RedirectAttributes;
+
+@Controller
+@RequestMapping("/user")
```
Comment (akshaym-webonise) [inline src/main/java/com/webonise/rbs/controller/UserController.java:18]: controller mapping must be plural

File: src/main/java/com/webonise/rbs/service/UserServiceImpl.java (hunk 6 above)
Comment (akshaym-webonise) [reply src/main/java/com/webonise/rbs/service/UserServiceImpl.java:16]: did it go unnoticed?

**Review by akshaym-webonise — APPROVED** [review]
Merging it for now. @antriksh-webonise Make sure you work on these comments in next pull request.

File: src/main/java/com/webonise/rbs/controller/UserController.java [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+    @PutMapping
+    public String updateUser(@ModelAttribute("editUser") User editUser, final RedirectAttributes redirectAttributes) {
+        if (userService.editUser(editUser) != null) {
+            redirectAttributes.addFlashAttribute("edit", "success");
+        } else {
+            redirectAttributes.addFlashAttribute("edit", "failure");
+        }
+        return "redirect:/user";
+    }
+}
```
Comment (himani-webonise) [inline src/main/java/com/webonise/rbs/controller/UserController.java:72]: Add new line at the EOF. Do this in other files as well
http://stackoverflow.com/questions/5813311/no-newline-at-end-of-file

**Pull request merged**